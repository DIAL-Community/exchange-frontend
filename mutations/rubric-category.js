import { gql } from '@apollo/client'

export const CREATE_RUBRIC_CATEGORY = gql`
  mutation CreateRubricCategory (
    $name: String!
    $slug: String!
    $weight: Float!
    $description: String!
  ) {
    createRubricCategory(
      name: $name
      slug: $slug
      weight: $weight
      description: $description
    ) {
      rubricCategory{
        id
        slug
      }
      errors
    }
  }
`

export const DELETE_RUBRIC_CATEGORY = gql`
  mutation DeleteRubricCategory ($id: ID!) {
    deleteRubricCategory(id: $id) {
      errors
    }
  }
`

export const UPDATE_RUBRIC_CATEGORY_INDICATORS = gql`
  mutation UpdateRubricCategoryIndicators (
    $categoryIndicatorSlugs: [String!]!
    $rubricCategorySlug: String!
  ) {
    updateRubricCategoryIndicators (
      categoryIndicatorSlugs: $categoryIndicatorSlugs
      rubricCategorySlug: $rubricCategorySlug
    ) {
      rubricCategory {
        categoryIndicators {
          slug
          name
          rubricCategoryId
        }
      } 
      errors
    }
  }
`
