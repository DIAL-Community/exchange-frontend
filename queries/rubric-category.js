import { gql } from '@apollo/client'

export const RUBRIC_CATEGORIES_LIST_QUERY = gql`
  query RubricCategories($search: String!) { 
    rubricCategories(search: $search) {
      nodes {
        name
        slug
        weight
      }
      totalCount
    }
  }
`

export const RUBRIC_CATEGORY_QUERY = gql`
  query RubricCategory ($slug: String!) {
    rubricCategory (slug: $slug) {
      name
      slug
      weight
      rubricCategoryDescription {
        description
      }
    }
  }
`
