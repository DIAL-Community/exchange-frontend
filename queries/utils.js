const MIN_SEARCH_VALUE_CHARACTERS = 2

export const fetchSelectOptions = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}

export const fetchSelectOptionsWithMature = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input,
    mature: true
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}
