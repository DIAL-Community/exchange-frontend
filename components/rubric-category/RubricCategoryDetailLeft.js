import Bookmark from '../shared/common/Bookmark'
import Comment from '../shared/common/Comment'
import Share from '../shared/common/Share'
import { ObjectType } from '../utils/constants'
import RubricCategoryDetailHeader from './fragments/RubricCategoryDetailHeader'
import RubricCategoryDetailNav from './fragments/RubricCategoryDetailNav'

const RubricCategoryDetailLeft = ({ scrollRef, rubricCategory }) => {
  return (
    <div className='bg-dial-slate-100 lg:h-full'>
      <div className='flex flex-col gap-y-3 px-4 lg:px-6 lg:py-3'>
        <RubricCategoryDetailHeader rubricCategory={rubricCategory} />
        <hr className='border-b border-dial-slate-200' />
        <RubricCategoryDetailNav rubricCategory={rubricCategory} scrollRef={scrollRef} />
        <hr className='border-b border-dial-slate-200' />
        <div className='hidden lg:flex flex-col gap-y-3'>
          <Bookmark object={rubricCategory} objectType={ObjectType.SDG} />
          <hr className='border-b border-dial-slate-200' />
          <Share />
          <hr className='border-b border-dial-slate-200' />
          <Comment entityKey={'ui.rubricCategory.label'} scrollRef={scrollRef} />
          <hr className='border-b border-dial-slate-200' />
        </div>
      </div>
    </div>
  )
}

export default RubricCategoryDetailLeft
