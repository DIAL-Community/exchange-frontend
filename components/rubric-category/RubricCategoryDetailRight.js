import { forwardRef, useCallback, useImperativeHandle, useRef } from 'react'
import { useIntl } from 'react-intl'
import { useQuery } from '@apollo/client'
import { EDITING_POLICY_SLUG, GRAPH_QUERY_CONTEXT } from '../../lib/apolloClient'
import CategoryIndicatorCard from '../category-indicator/CategoryIndicatorCard'
import CommentsSection from '../shared/comment/CommentsSection'
import Bookmark from '../shared/common/Bookmark'
import Share from '../shared/common/Share'
import CreateButton from '../shared/form/CreateButton'
import EditButton from '../shared/form/EditButton'
import { HtmlViewer } from '../shared/form/HtmlViewer'
import { RUBRIC_CATEGORY_QUERY } from '../shared/query/rubricCategory'
import { DisplayType, ObjectType } from '../utils/constants'
import DeleteRubricCategory from './buttons/DeleteRubricCategory'

const RubricCategoryDetailRight = forwardRef(({ rubricCategory }, ref) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const descRef = useRef()
  const categoryIndicatorRef = useRef()
  const commentsSectionRef = useRef()
  useImperativeHandle(
    ref,
    () => [
      { value: 'ui.common.detail.description', ref: descRef },
      { value: 'categoryIndicator.header', ref: categoryIndicatorRef },
      { value: 'ui.comment.label', ref: commentsSectionRef }
    ],
    []
  )

  const editPath = `${rubricCategory.slug}/edit`

  let editingAllowed = true
  const { error } = useQuery(RUBRIC_CATEGORY_QUERY, {
    variables: { slug: EDITING_POLICY_SLUG },
    context: {
      headers: {
        ...GRAPH_QUERY_CONTEXT.EDITING
      }
    }
  })

  if (error) {
    editingAllowed = false
  }

  return (
    <div className='px-4 lg:px-0 py-4 lg:py-6'>
      <div className='flex flex-col gap-y-3'>
        {editingAllowed && (
          <div className='flex gap-x-3 ml-auto'>
            <EditButton type='link' href={editPath} />
            <DeleteRubricCategory rubricCategory={rubricCategory} />
          </div>
        )}
        <div className='text-xl font-semibold text-dial-plum py-3' ref={descRef}>
          {format('ui.common.detail.description')}
        </div>
        <div className='block'>
          <HtmlViewer
            initialContent={rubricCategory?.rubricCategoryDescription?.description}
            editorId='rubric-category-description'
          />
        </div>
        <hr className='border-b border-dial-blue-chalk my-3'/>
        <div className='flex flex-col gap-y-3'>
          <div className='flex flex-row pb-3' ref={categoryIndicatorRef}>
            <div className='text-xl font-semibold text-dial-blueberry '>
              {format('categoryIndicator.header')}
            </div>
            {editingAllowed &&
              <CreateButton
                type='link'
                className='ml-auto'
                label={format('app.create')}
                href={
                  `/rubric-categories/${rubricCategory.slug}` +
                  '/category-indicators/create'
                }
              />
            }
          </div>
          <div className='flex flex-col gap-y-3'>
            {rubricCategory?.categoryIndicators?.map((categoryIndicator, index) =>
              <div key={index}>
                <CategoryIndicatorCard
                  rubricCategory={rubricCategory}
                  categoryIndicator={categoryIndicator}
                  displayType={DisplayType.SMALL_CARD}
                />
              </div>
            )}
          </div>
        </div>
        <hr className='border-b border-dial-blue-chalk my-3' />
        <div className='lg:hidden flex flex-col gap-y-3'>
          <Bookmark object={rubricCategory} objectType={ObjectType.RUBRIC_CATEGORY} />
          <hr className='border-b border-dial-slate-200'/>
          <Share />
          <hr className='border-b border-dial-slate-200'/>
        </div>
        <CommentsSection
          commentsSectionRef={commentsSectionRef}
          objectId={rubricCategory.id}
          objectType={ObjectType.RUBRIC_CATEGORY}
        />
      </div>
    </div>
  )
})

RubricCategoryDetailRight.displayName = 'RubricCategoryDetailRight'

export default RubricCategoryDetailRight
