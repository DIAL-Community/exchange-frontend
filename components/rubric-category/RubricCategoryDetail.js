import { useEffect, useRef, useState } from 'react'
import { useApolloClient, useQuery } from '@apollo/client'
import { GRAPH_QUERY_CONTEXT } from '../../lib/apolloClient'
import Breadcrumb from '../shared/Breadcrumb'
import { handleLoadingQuery, handleMissingData, handleQueryError } from '../shared/GraphQueryHandler'
import { RUBRIC_CATEGORY_POLICY_QUERY, RUBRIC_CATEGORY_QUERY } from '../shared/query/rubricCategory'
import { fetchOperationPolicies } from '../utils/policy'
import RubricCategoryDetailLeft from './RubricCategoryDetailLeft'
import RubricCategoryDetailRight from './RubricCategoryDetailRight'

const RubricCategoryDetail = ({ categorySlug }) => {
  const scrollRef = useRef(null)
  const client = useApolloClient()

  const [editingAllowed, setEditingAllowed] = useState(false)
  const [deletingAllowed, setDeletingAllowed] = useState(false)

  const { loading, error, data } = useQuery(RUBRIC_CATEGORY_QUERY, {
    variables: { slug: categorySlug },
    context: {
      headers: {
        ...GRAPH_QUERY_CONTEXT.VIEWING
      }
    }
  })

  useEffect(() => {
    fetchOperationPolicies(
      client,
      RUBRIC_CATEGORY_POLICY_QUERY,
      ['editing', 'deleting']
    ).then(policies => {
      setEditingAllowed(policies['editing'])
      setDeletingAllowed(policies['deleting'])
    })
  }, [client])

  if (loading) {
    return handleLoadingQuery()
  } else if (error) {
    return handleQueryError(error)
  } else if (!data?.rubricCategory) {
    return handleMissingData()
  }

  const { rubricCategory } = data

  const slugNameMapping = (() => {
    const map = {}
    map[rubricCategory.slug] = rubricCategory.name

    return map
  })()

  return (
    <div className='lg:px-8 xl:px-56 flex flex-col'>
      <div className='px-4 lg:px-6 py-4 bg-dial-violet text-dial-stratos ribbon-detail z-40'>
        <Breadcrumb slugNameMapping={slugNameMapping}/>
      </div>
      <div className='flex flex-col lg:flex-row gap-x-8'>
        <div className='lg:basis-1/3 shrink-0'>
          <RubricCategoryDetailLeft
            scrollRef={scrollRef}
            rubricCategory={rubricCategory}
            editingAllowed={editingAllowed}
            deletingAllowed={deletingAllowed}
          />
        </div>
        <div className='lg:basis-2/3 shrink-0'>
          <RubricCategoryDetailRight
            ref={scrollRef}
            rubricCategory={rubricCategory}
            editingAllowed={editingAllowed}
            deletingAllowed={deletingAllowed}
          />
        </div>
      </div>
    </div>
  )
}

export default RubricCategoryDetail
