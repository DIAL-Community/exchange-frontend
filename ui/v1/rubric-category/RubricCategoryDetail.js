import { useRef } from 'react'
import { useQuery } from '@apollo/client'
import { RUBRIC_CATEGORY_QUERY } from '../shared/query/rubricCategory'
import Breadcrumb from '../shared/Breadcrumb'
import { Error, Loading, NotFound } from '../shared/FetchStatus'
import RubricCategoryDetailRight from './RubricCategoryDetailRight'
import RubricCategoryDetailLeft from './RubricCategoryDetailLeft'

const RubricCategoryDetail = ({ categorySlug }) => {
  const scrollRef = useRef(null)

  const { loading, error, data } = useQuery(RUBRIC_CATEGORY_QUERY, {
    variables: { slug: categorySlug }
  })

  if (loading) {
    return <Loading />
  } else if (error) {
    return <Error />
  } else if (!data?.rubricCategory) {
    return <NotFound />
  }

  const { rubricCategory } = data

  const slugNameMapping = (() => {
    const map = {}
    map[rubricCategory.slug] = rubricCategory.name

    return map
  })()

  return (
    <div className='lg:px-8 xl:px-56 flex flex-col'>
      <div className='px-4 lg:px-6 py-4 bg-dial-violet text-dial-stratos ribbon-detail z-40'>
        <Breadcrumb slugNameMapping={slugNameMapping}/>
      </div>
      <div className='flex flex-col lg:flex-row gap-x-8'>
        <div className='lg:basis-1/3'>
          <RubricCategoryDetailLeft scrollRef={scrollRef} rubricCategory={rubricCategory} />
        </div>
        <div className='lg:basis-2/3'>
          <RubricCategoryDetailRight ref={scrollRef} rubricCategory={rubricCategory} />
        </div>
      </div>
    </div>
  )
}

export default RubricCategoryDetail
