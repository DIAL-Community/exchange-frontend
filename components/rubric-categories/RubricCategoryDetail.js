import { useQuery } from '@apollo/client'
import { useEffect, useMemo } from 'react'
import NotFound from '../shared/NotFound'
import { Error, Loading } from '../shared/FetchStatus'
import { RUBRIC_CATEGORY_QUERY } from '../../queries/rubric-category'
import RubricCategoryDetailLeft from './RubricCategoryDetailLeft'
import RubricCategoryDetailRight from './RubricCategoryDetailRight'

const RubricCategoryDetail = ({ slug, locale }) => {
  const { loading, error, data, refetch } = useQuery(RUBRIC_CATEGORY_QUERY, {
    variables: { slug },
    context: { headers: { 'Accept-Language': locale } },
  })

  useEffect(
    () => {
      refetch()
    },
    [refetch, locale]
  )

  const slugNameMapping = useMemo(() => ({ [data?.rubricCategory.slug]: data?.rubricCategory.name }), [data])

  if (loading) {
    return <Loading />
  } else if (error && error.networkError) {
    return <Error />
  } else if (error && !error.networkError) {
    return <NotFound />
  }

  return (
    <div className='flex flex-col lg:flex-row pb-8 max-w-catalog mx-auto'>
      <div className='lg:sticky lg:top-66px w-full lg:w-1/3 xl:w-1/4 h-full p-4'>
        <RubricCategoryDetailLeft rubricCategory={data?.rubricCategory} slugNameMapping={slugNameMapping} />
      </div>
      <div className='w-full lg:w-2/3 xl:w-3/4'>
        <RubricCategoryDetailRight rubricCategory={data?.rubricCategory} slugNameMapping={slugNameMapping} />
      </div>
    </div>
  )
}

export default RubricCategoryDetail
