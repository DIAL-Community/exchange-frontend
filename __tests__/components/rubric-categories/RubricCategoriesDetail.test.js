import { render, waitForAllEffects } from '../../test-utils'
import CustomMockedProvider, { generateMockApolloData } from '../../utils/CustomMockedProvider'
import RubricCategoryDetail from '../../../components/rubric-categories/RubricCategoryDetail'
import { RUBRIC_CATEGORY_QUERY } from '../../../queries/rubric-category'
import { mockNextAuthUseSession, mockNextUseRouter, statuses } from '../../utils/nextMockImplementation'
import { rubricCategory } from './data/RubricCategoryDetail'

mockNextUseRouter()
describe('Unit test for the RubricCategoryDetail component.', () => {
  const mockRubricCategory = generateMockApolloData(RUBRIC_CATEGORY_QUERY, { slug: 'rc_name' }, null, rubricCategory)

  test('Should render RubricCategoryDetail component', async () => {
    mockNextAuthUseSession(statuses.AUTHENTICATED, { canEdit: true })
    const { container, getByTestId } = render(
      <CustomMockedProvider mocks={[mockRubricCategory]}>
        <RubricCategoryDetail slug='rc_name'/>
      </CustomMockedProvider>
    )
    await waitForAllEffects()

    expect(getByTestId('description')).toHaveTextContent('RC description')

    expect(container).toMatchSnapshot()
  })
})
