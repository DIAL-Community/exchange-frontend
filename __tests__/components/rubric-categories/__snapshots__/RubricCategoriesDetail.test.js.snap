// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for the RubricCategoryDetail component. Should render RubricCategoryDetail component 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="flex flex-col lg:flex-row pb-8 max-w-catalog mx-auto"
  >
    <div
      class="lg:sticky lg:top-66px w-full lg:w-1/3 xl:w-1/4 h-full p-4"
    >
      <div
        class="block lg:hidden"
      >
        <div
          class="bg-white pb-3 lg:py-4 whitespace-nowrap text-ellipsis overflow-hidden"
        >
          <a
            class="inline text-dial-blue h5"
            href="/"
          >
            Home
          </a>
          <div
            class="inline h5"
          >
             &gt; 
            <a
              class="text-dial-gray-dark"
              href="/"
            />
          </div>
        </div>
      </div>
      <div
        class="h-12 w-full"
      />
      <div
        class="bg-white border-2 border-dial-gray lg:mr-6 shadow-lg"
      >
        <div
          class="flex flex-col p-4 text-dial-gray-dark"
        >
          <div
            class="font-semibold text-lg"
          >
            RC name
          </div>
          <hr
            class="my-2"
          />
          <div
            class="text-sm"
          >
            <span
              class="font-semibold"
            >
              Weight
            </span>
            : 1
          </div>
        </div>
      </div>
    </div>
    <div
      class="w-full lg:w-2/3 xl:w-3/4"
    >
      <div
        class="px-4"
      >
        <div
          class="hidden lg:block"
        >
          <div
            class="bg-white pb-3 lg:py-4 whitespace-nowrap text-ellipsis overflow-hidden"
          >
            <a
              class="inline text-dial-blue h5"
              href="/"
            >
              Home
            </a>
            <div
              class="inline h5"
            >
               &gt; 
              <a
                class="text-dial-gray-dark"
                href="/"
              />
            </div>
          </div>
        </div>
        <div
          class="mt-8 card-title mb-3 text-dial-gray-dark"
        >
          Description
        </div>
        <div
          class="fr-view text-dial-gray-dark"
          data-testid="description"
        >
          RC description
        </div>
        <div
          class="mt-8 card-title mb-3 text-dial-gray-dark"
        >
          Indicators
        </div>
        <div
          class="grid sm:grid-cols-1 lg:grid-cols-2"
          data-testid="indicators"
        >
          <div
            class="border-3 border-transparent hover:border-dial-yellow cursor-pointer"
          >
            <div
              class="font-semibold text-button-gray items-center grid gap-x-4 py-4 px-8 w-full border border-dial-gray card-drop-shadow"
            >
              CI name
            </div>
          </div>
          <div
            class="border-3 border-transparent hover:border-dial-yellow cursor-pointer"
          >
            <div
              class="font-semibold text-button-gray items-center grid gap-x-4 py-4 px-8 w-full border border-dial-gray card-drop-shadow"
            >
              CI name 2
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
